src/main/java/com/icthh/xm/tmf/ms/activation/events/KafkaEventsSender.java

```diff
@@ -37,20 +40,31 @@ public class KafkaEventsSender implements EventsSender {
                multiplierExpression = "${application.kafkaEventSender.retry.multiplier}"))
     @Override
     public void sendEvent(SagaEvent sagaEvent) {
-        boolean result = channelResolver
-            .resolveDestination(queueNameResolver.resolveQueueName(sagaEvent))
-            .send(MessageBuilder.withPayload(sagaEvent)
-                                .setHeader(KafkaHeaders.MESSAGE_KEY, sagaEvent.getId())
-                                .setHeader(PARTITION_KEY, self.getPartitionKey(sagaEvent))
-                                .build());
+        String queueName = queueNameResolver.resolveQueueName(sagaEvent);
+        Integer partitionKey = getPartitionKeyForTopic(queueName, sagaEvent);
+        String payload = getMessagePayload(sagaEvent);
+
+        template.send(queueName, partitionKey, sagaEvent.getId(), payload);
 
-        if (!result) {
-            log.warn("Cannot send saga event: {}", sagaEvent);
-            throw new BusinessException("Cannot send saga event: " + sagaEvent);
-        }
         log.info("Saga event successfully sent: {}", sagaEvent);
     }
 
+    private String getMessagePayload(SagaEvent sagaEvent) {
+        try {
+            String sagaEventJson = objectMapper.writeValueAsString(sagaEvent);
+            return Base64.getEncoder().encodeToString(sagaEventJson.getBytes(UTF_8));
+
+        } catch (JsonProcessingException e) {
+            log.warn("Cannot convert message to json: {}", sagaEvent, e);
+            throw new BusinessException("Cannot convert message to json: " + sagaEvent);
+        }
+    }
+
+    public Integer getPartitionKeyForTopic(String topic, SagaEvent sagaEvent) {
+        int partitionCount = template.partitionsFor(topic).size();
+        return Math.abs(getPartitionKey(sagaEvent).hashCode()) % partitionCount;
```

Comment:
direct call getPartitionKey will not invoke lep GetPartitionKey

Reply:
done